File: Server_demo/RPC/src/main/java/com/RPC/Serializer.java
package com.RPC;

// Serializer.java
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class Serializer {

    // 反序列化请求
    public static RPCRequest unmarshallRequest(byte[] data, int length) throws Exception {
        ByteBuffer buffer = ByteBuffer.wrap(data, 0, length);

        // 方法名长度
        if (buffer.remaining() < 1) throw new Exception("Invalid request: Missing method name length.");
        int methodNameLen = Byte.toUnsignedInt(buffer.get());

        // 方法名
        if (buffer.remaining() < methodNameLen) throw new Exception("Invalid request: Incomplete method name.");
        byte[] methodNameBytes = new byte[methodNameLen];
        buffer.get(methodNameBytes);
        String methodName = new String(methodNameBytes, StandardCharsets.UTF_8);

        // 参数数量
        if (buffer.remaining() < 1) throw new Exception("Invalid request: Missing parameter count.");
        int paramCount = Byte.toUnsignedInt(buffer.get());

        List<Parameter> parameters = new ArrayList<>();
        for (int i = 0; i < paramCount; i++) {
            // 参数类型
            if (buffer.remaining() < 1) throw new Exception("Invalid request: Missing parameter type.");
            int typeCode = Byte.toUnsignedInt(buffer.get());
            ParameterType type = ParameterType.fromCode(typeCode);

            // 参数值
            switch (type) {
                case INTEGER:
                    if (buffer.remaining() < 4) throw new Exception("Invalid request: Incomplete integer parameter.");
                    int intValue = buffer.getInt();
                    parameters.add(new Parameter(type, intValue));
                    break;
                case STRING:
                    if (buffer.remaining() < 1) throw new Exception("Invalid request: Missing string length.");
                    int strLen = Byte.toUnsignedInt(buffer.get());
                    if (buffer.remaining() < strLen) throw new Exception("Invalid request: Incomplete string parameter.");
                    byte[] strBytes = new byte[strLen];
                    buffer.get(strBytes);
                    String strValue = new String(strBytes, StandardCharsets.UTF_8);
                    parameters.add(new Parameter(type, strValue));
                    break;
                // 可以根据需要添加更多类型的处理
                default:
                    throw new Exception("Unsupported parameter type: " + type);
            }
        }

        return new RPCRequest(methodName, parameters);
    }

    // 序列化响应
    public static byte[] marshallResponse(RPCResponse response) throws Exception {
        List<Byte> bytes = new ArrayList<>();
        bytes.add(response.getStatus());

        if (response.getStatus() == 0) { // success
            List<Parameter> results = response.getResults();
            if (results.size() > 255) {
                throw new Exception("Too many result parameters to serialize.");
            }
            bytes.add((byte) results.size());

            for (Parameter result : results) {
                bytes.add((byte) result.getType().getCode());
                switch (result.getType()) {
                    case INTEGER:
                        int intValue = (Integer) result.getValue();
                        ByteBuffer intBuffer = ByteBuffer.allocate(4);
                        intBuffer.putInt(intValue);
                        for (byte b : intBuffer.array()) {
                            bytes.add(b);
                        }
                        break;
                    case STRING:
                        String strValue = (String) result.getValue();
                        byte[] strBytes = strValue.getBytes(StandardCharsets.UTF_8);
                        if (strBytes.length > 255) {
                            throw new Exception("String result too long to serialize.");
                        }
                        bytes.add((byte) strBytes.length);
                        for (byte b : strBytes) {
                            bytes.add(b);
                        }
                        break;
                    // 可以根据需要添加更多类型的处理
                    default:
                        throw new Exception("Unsupported result type: " + result.getType());
                }
            }
        } else { // error
            String errorMsg = response.getErrorMessage();
            byte[] errorBytes = errorMsg.getBytes(StandardCharsets.UTF_8);
            if (errorBytes.length > 255) {
                errorBytes = "Error message too long.".getBytes(StandardCharsets.UTF_8);
            }
            bytes.add((byte) errorBytes.length);
            for (byte b : errorBytes) {
                bytes.add(b);
            }
        }

        // 转换为byte数组
        byte[] byteArray = new byte[bytes.size()];
        for (int i = 0; i < bytes.size(); i++) {
            byteArray[i] = bytes.get(i);
        }
        return byteArray;
    }
}

File: Client_C++/ConsoleApplication1/ConsoleApplication1/Serializer.cpp
#include "Serializer.h"
#include <cstring>
#include <stdexcept>
#include <winsock2.h>
#include <ws2tcpip.h>

// 确保 Winsock 被正确初始化
#pragma comment(lib, "Ws2_32.lib")




// convert float to network byte order (big endian)
uint32_t convfton(float f) {
    uint32_t p;
    std::memcpy(&p, &f, sizeof(float));
    // Convert to big endian
    return htonl(p);
}

// convert float from network byte order (big endian)
float convntof(uint32_t p) {
    p = ntohl(p);
    float f;
    std::memcpy(&f, &p, sizeof(float));
    return f;
}






// Marshalling: 序列化RPCRequest到字节数组
std::vector<char> marshallRequest(const RPCRequest& request) {
    std::vector<char> buffer;

    // 请求标识符
    int netRequestId = htonl(request.requestId);
    char* p = reinterpret_cast<char*>(&netRequestId);
    buffer.insert(buffer.end(), p, p + sizeof(int));


    // 方法名长度
    if (request.methodName.length() > 255) {
        throw std::runtime_error("Method name too long!");
    }
    buffer.push_back(static_cast<char>(request.methodName.length()));

    // 方法名
    buffer.insert(buffer.end(), request.methodName.begin(), request.methodName.end());

    // 参数数量
    if (request.parameters.size() > 255) {
        throw std::runtime_error("Too many parameters!");
    }
    buffer.push_back(static_cast<char>(request.parameters.size()));

    // 每个参数
    for (const auto& param : request.parameters) {
        // 参数类型
        buffer.push_back(static_cast<char>(param.type));

        // 参数值
        switch (param.type) {
        case INTEGER: {
            int netInt = htonl(param.intValue);
            char* p = reinterpret_cast<char*>(&netInt);
            buffer.insert(buffer.end(), p, p + sizeof(int));
            break;
        }
        case STRING: {
            std::string str = *(param.stringValue);
            if (str.length() > 255) {
                throw std::runtime_error("String parameter too long!");
            }
            buffer.push_back(static_cast<char>(str.length()));
            buffer.insert(buffer.end(), str.begin(), str.end());
            break;
        }
        case MYFLOAT: { 
            uint32_t netFloat = convfton(param.floatValue);
            char* p = reinterpret_cast<char*>(&netFloat);
            buffer.insert(buffer.end(), p, p + sizeof(uint32_t));
            break;
        }
        default:
            throw std::runtime_error("Unsupported parameter type!");
        }
    }

    return buffer;
}

// Unmarshalling: 反序列化字节数组到RPCResponse
RPCResponse unmarshallResponse(const char* data, int length) {
    RPCResponse response;
    int index = 0;

    if (length < 4) {
        throw std::runtime_error("Invalid response length: Missing request ID.");
    }

    // 请求标识符
    int netRequestId;
    std::memcpy(&netRequestId, data + index, sizeof(int));
    response.requestId = ntohl(netRequestId);
    index += sizeof(int);

    if (length < index + 1) {
        throw std::runtime_error("Invalid response length: Missing status.");
    }


   // if (length < 1) {
     //   throw std::runtime_error("Invalid response length.");
   // }

    // 状态码
    response.status = static_cast<uint8_t>(data[index]);
    index += 1;

    if (response.status == 0) { // success
        if (length < index + 1) {
            throw std::runtime_error("Invalid response length for success.");
        }

        // 结果参数数量
        int resultCount = static_cast<uint8_t>(data[index]);
        index += 1;

        for (int i = 0; i < resultCount; i++) {
            if (length < index + 1) {
                throw std::runtime_error("Invalid response length: Missing result parameter type.");
            }
            ParameterType resultType = static_cast<ParameterType>(static_cast<uint8_t>(data[index]));
            index += 1;

            switch (resultType) {
            case INTEGER: {
                if (length < index + 4) {
                    throw std::runtime_error("Invalid response length for integer result.");
                }
                int netInt;
                std::memcpy(&netInt, data + index, sizeof(int));
                int intValue = ntohl(netInt);
                response.results.emplace_back(intValue);
                index += 4;
                break;
            }
            case STRING: {
                if (length < index + 1) {
                    throw std::runtime_error("Invalid response length for string result.");
                }
                int strLen = static_cast<uint8_t>(data[index]);
                index += 1;
                if (length < index + strLen) {
                    throw std::runtime_error("Invalid response length for string result.");
                }
                std::string strValue(data + index, strLen);
                response.results.emplace_back(strValue);
                index += strLen;
                break;
            }
            case MYFLOAT: { 
                if (length < index + 4) {
                    throw std::runtime_error("Invalid response length for float result.");
                }
                uint32_t netFloat;
                std::memcpy(&netFloat, data + index, sizeof(uint32_t));
                float floatValue = convntof(netFloat);
                response.results.emplace_back(floatValue);
                index += 4;
                break;
            }
            default:
                throw std::runtime_error("Unsupported result type.");
            }
        }
    }
    else { // error
        if (length < index + 1) {
            throw std::runtime_error("Invalid response length for error.");
        }

        // 错误消息长度
        int errorMsgLen = static_cast<uint8_t>(data[index]);
        index += 1;

        if (length < index + errorMsgLen) {
            throw std::runtime_error("Invalid response length for error message.");
        }

        // 错误消息
        response.errorMessage = std::string(data + index, errorMsgLen);
        index += errorMsgLen;
    }

    return response;
}
